`rptest`: basic `avro` schema generation and `iceberg` testing

WillemKauf opened the pull request:
Adds a schema generation framework (WIP, currently targeting the `avro` spec, no `protobuf` code yet) and a `schema_scale_test` that simply produces to an `iceberg` enabled topic and waits for translation to occur. 

The randomly generated schema can be controlled in terms of maximum depth, number of fields, and even the likeliness of primitive/complex data types in the schema. A template record is constructed that allows for the random generation of a schema-compatible record for producing to `redpanda`. This record is verified using the schema before being produced.

Parts of the random schema generation framework are still quite a work in progress- not all complex data types are supported, and the implementation (which hopes to be generic) is fairly hardcoded for the `avro` spec currently.

Future work entails teasing the framework apart to de-couple the generic bits away from a specific format, and add new abstraction layers on top for `protobuf` and `avro` respectively.


## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

WillemKauf committed:
`rptest`: add `data_types.py`

Various definitions for generic representations of common types for
use in protobuf/avro schema generation.

WillemKauf committed:
`rptest`: add `schema_scale_test`

For testing randomly generated schemas and their effect
on `redpanda/iceberg`.

rockwotj commented:
FYI if it's helpful there is some code in `datalake_e2e_test.py` that can generate protocol buffer classes dynamically in Python.

<details><summary>Code inlined below if it's helpful</summary>
<p>

```
from google import protobuf
from google.protobuf import text_format as pb_text_format, json_format as pb_json_format

        protobuf_schema_v1 = """
        syntax = "proto3";
        message Person {
          string name = 1;
          int32 id = 2;
          string email = 3;
        }
        """
        protobuf_schema_v2 = """
        syntax = "proto3";
        message Address {
          string street = 1;
          string city = 2;
          string state = 3;
          string zip = 4;
        }
        message Person {
          string name = 1;
          int32 id = 2;
          string email = 3;
          Address address = 4;  // Nested message
        }
        """
        # The text format of the above protobuf as to not have to bother setting up protoc for this test.
        protobuf_file_descriptor = """
        name: "example.proto"
        message_type {
          name: "Address"
          field { name: "street" number: 1 label: LABEL_OPTIONAL type: TYPE_STRING }
          field { name: "city" number: 2 label: LABEL_OPTIONAL type: TYPE_STRING }
          field { name: "state" number: 3 label: LABEL_OPTIONAL type: TYPE_STRING }
          field { name: "zip" number: 4 label: LABEL_OPTIONAL type: TYPE_STRING }
        }
        message_type {
          name: "Person"
          field { name: "name" number: 1 label: LABEL_OPTIONAL type: TYPE_STRING }
          field { name: "id" number: 2 label: LABEL_OPTIONAL type: TYPE_INT32 }
          field { name: "email" number: 3 label: LABEL_OPTIONAL type: TYPE_STRING }
          field { name: "address" number: 4 label: LABEL_OPTIONAL type: TYPE_MESSAGE type_name: ".Address" }
        }
        """

        # Using the protobuf text format, parse the raw descriptor and create a dynamic message from it.
        file_desc_pb = protobuf.descriptor_pb2.FileDescriptorProto()
        pb_text_format.Merge(protobuf_file_descriptor, file_desc_pb)
        pool = protobuf.descriptor_pool.DescriptorPool()
        pool.Add(file_desc_pb)
        factory = protobuf.message_factory.MessageFactory(pool)
        person_desc = pool.FindMessageTypeByName("Person")
        Person = factory.GetPrototype(person_desc)
        def produce_protos():
            producer = Producer({'bootstrap.servers': self.redpanda.brokers()})
            for i in range(count):
                record = json.dumps({
                    "name": f"Bob{i} Protopants",
                    "id": 1 + i,
                    "email": f"foobar{i}@gmail.com",
                    "address": {
                        "street": f"{i} Main St.",
                        "city": "Protoville" if i % 2 == 0 else "Buftown",
                        "state": "District 13" if i % 3 == 0 else "Hooli",
                        "zip": "8675309" if i % 4 == 0 else "12345"
                    }
                })
                person_proto = Person()
                pb_json_format.Parse(record, person_proto)
                producer.produce(topic=self.topic_name,
                                 value=person_proto.SerializeToString())
            producer.flush()
```

</p>
</details>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--803116f5-394c-4f76-a0c1-64094dc96276-->
<!--TEST RESULTS-->
<!--dc301d9d-717a-4347-bd70-7b5e5b0f89c5-->
<!--TEST RESULTS-->
<!--33acf53f-d086-4aad-b38b-39f41495a193-->
<!--TEST RESULTS-->
<!--228f96c5-ccea-46d6-a308-0be9a90e6894-->
<!--TEST RESULTS-->
<!--5b769d81-8588-47e5-89e9-3e0248941cb3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63690">build#63690</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5510-44fd-842d-511631a9a440 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d3-4168-aa5f-4f32c45d33b5 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5514-44af-bc7c-44ab47f518d2 | FLAKY | 1/7 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d1-46ff-85be-23ecd1fef02f | FLAKY | 1/2 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5511-4e91-946f-42dc61fcba9c | FLAKY | 1/4 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d3-4168-aa5f-4f32c45d33b5 | FLAKY | 1/8 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5514-44af-bc7c-44ab47f518d2 | FLAKY | 1/14 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d1-46ff-85be-23ecd1fef02f | FLAKY | 1/3 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5511-4e91-946f-42dc61fcba9c | FLAKY | 1/5 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d1-46ff-85be-23ecd1fef02f | FLAKY | 1/6 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5511-4e91-946f-42dc61fcba9c | FLAKY | 1/13 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d1-46ff-85be-23ecd1fef02f | FLAKY | 1/2 |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path.test_case=.TS_Read==True.TS_Timequery==True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1c-5511-4e91-946f-42dc61fcba9c | FLAKY | 1/2 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d1-46ff-85be-23ecd1fef02f | FLAKY | 1/2 |
| rptest.tests.topic_delete_test.TopicDeleteCloudStorageTest.topic_delete_installed_snapshots_test | ducktape | https://buildkite.com/redpanda/redpanda/builds/63690#0195cf1f-b6d3-4e14-8cde-fa753192a9ac | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63894">build#63894</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec4-495d-aa0b-6c29ab9306d8 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec6-4405-88a3-8a79de100183 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec4-4af1-8491-69c4d3b8b772 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea9-3b7b-45be-8023-0213d8d4fa71 | FAIL | 0/1 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec6-4405-88a3-8a79de100183 | FLAKY | 1/7 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea9-3b79-4cf4-9595-35c73ced5aad | FAIL | 0/1 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea9-3b7b-45be-8023-0213d8d4fa71 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec6-4405-88a3-8a79de100183 | FLAKY | 1/4 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea9-3b79-4cf4-9595-35c73ced5aad | FAIL | 0/1 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec4-495d-aa0b-6c29ab9306d8 | FLAKY | 1/10 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec6-4405-88a3-8a79de100183 | FLAKY | 1/6 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea9-3b7b-45be-8023-0213d8d4fa71 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/63894#0195dea5-8ec4-495d-aa0b-6c29ab9306d8 | FLAKY | 1/3 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63894#0195de64-0850-4430-b74b-13c56d63e00d | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63988">build#63988</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195edeb-23b1-48be-b687-d5630bb6345e | FAIL | 0/1 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195ede8-4672-44de-86b4-7f7f5d15b4c6 | FLAKY | 1/5 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.NESSIE.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195ede8-4674-402a-8cab-aa0277d13794 | FLAKY | 1/5 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_HADOOP.cleanup_policy=compact.use_partition_spec=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195edeb-23b2-4920-9439-bd3102153ea7 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_scale_test.SchemaScaleTest.schema_scale_test.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_JDBC.cleanup_policy=compact.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195ede8-4674-402a-8cab-aa0277d13794 | FLAKY | 1/6 |
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionLargeClusterTest.test_consumer_offset_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195ede8-4672-44de-86b4-7f7f5d15b4c6 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195edeb-23b1-48be-b687-d5630bb6345e | FLAKY | 1/5 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group | ducktape | https://buildkite.com/redpanda/redpanda/builds/63988#0195ede8-4673-4731-a634-ce573fd68d09 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64114">build#64114</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/64114#0195f73b-0913-47f4-b1fa-ce2128048a56 | FLAKY | 1/2 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/64114#0195f73d-74b0-4c1a-a1a7-0ed591ebe803 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64132">build#64132</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/64132#0195f823-efd9-4eb1-8025-18f67648eed7 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64132#0195f881-5194-413e-ade1-90d8a428f544 | FLAKY         | 1/2      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64132#0195f881-5198-4614-b4fa-b38ff13cf04b | FLAKY         | 1/2      |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64132#0195f881-5194-413e-ade1-90d8a428f544 | FLAKY         | 1/2      |
</details>
<!--/body-->

WillemKauf commented:
Thanks @rockwotj, that will definitely streamline the production side of things when I get to the `protobuf` impl.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63894-->
<!--2431aa40-b5d7-45ac-8b20-5dad06b34a32-->
<!--015d8f7c-b041-422b-8a89-99654e4d875c-->
<!--title-->
## Retry command for [Build#63894](https://buildkite.com/redpanda/redpanda/builds/63894)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/schema_scale_test.py::SchemaScaleTest.schema_scale_test@{"catalog_type":"rest_hadoop","cleanup_policy":"compact","cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
tests/rptest/tests/datalake/schema_scale_test.py::SchemaScaleTest.schema_scale_test@{"catalog_type":"nessie","cleanup_policy":"compact","cloud_storage_type":1,"query_engine":"trino","use_partition_spec":true}
tests/rptest/tests/datalake/schema_scale_test.py::SchemaScaleTest.schema_scale_test@{"catalog_type":"nessie","cleanup_policy":"compact","cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
```
<!--/body-->

dotnwat commented on tests/rptest/tests/datalake/schemas/data_types.py:
```diff
@@ -0,0 +1,336 @@
+                           k=n))
+
+
+ALL_PRIMITIVE_DATA_TYPES = [
+    GenericBool, GenericInt, GenericLong, GenericFloat, GenericDouble,
+    GenericString
+]
+
+ALL_PRIMITIVE_DATA_TYPES_AND_ENUM = ALL_PRIMITIVE_DATA_TYPES + [GenericEnum]
+
+AVRO_PRIMITIVE_DATA_TYPES_RANDOM_GENS = {
+    gen_type.to_avro(): gen_type.random
+    for gen_type in ALL_PRIMITIVE_DATA_TYPES_AND_ENUM
+}
```
this is cool. it's all written now so no reason to change it, but i do wonder if you could have consolidated all of this into a raw dictionary (more compact representation) and then dynamically generated the classes.

dotnwat commented on tests/rptest/tests/datalake/schema_scale_test.py:
```diff
@@ -0,0 +1,189 @@
+                              redpanda=self.redpanda,
+                              include_query_engines=[query_engine],
+                              catalog_type=catalog_type) as dl:
+            self._make_schema(use_partition_spec)
```
am i understanding correctly that one instance of this test corresponds to one random schema? if that's the case, then would it make sense to take advantage of the cost of spinning up the test to generate a bunch of random schemas?

WillemKauf replied:
Modified this to run with 5 randomly generated schemas per test.

I'm curious as to whether this test is more beneficial if:
* We generate a ton of random schemas and simply test they work by producing a small amount of records to them, testing for any edge cases that may fail
* We generate a small amount of random schemas and stress the system by producing a large number of records to them, stressing the translation path and the broker as a whole

WDYT? Maybe it's worth further parameterizing the test further to cover both these angles.

andrwng replied:
>further parameterizing the test 

IMO these sound like two different test cases

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63988-->
<!--82ca5a24-58cd-4029-b9fb-d84cbf7377ce-->
<!--title-->
## Retry command for [Build#63988](https://buildkite.com/redpanda/redpanda/builds/63988)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/schema_scale_test.py::SchemaScaleTest.schema_scale_test@{"catalog_type":"rest_hadoop","cleanup_policy":"compact","cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
```
<!--/body-->

WillemKauf commented:
Force push to:
* Disable `use_partition_spec` for the time being until the OOM on high cardinality fields is fixed. The `schema_scale_test` written is useful on its own without this functionality for now.
* Modify `schema_scale_test` to run with five randomly generated schemas instead of just one.
* Enable `map` types, as they are now fixed in the `redpanda` iceberg implementation.

WillemKauf commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=5
tests/rptest/tests/datalake/schema_scale_test.py

andrwng commented on tests/rptest/tests/datalake/schema_scale_test.py:
```diff
@@ -0,0 +1,199 @@
+            dl.wait_for_translation(topic_name, msg_count=self.total + count)
+            self.total += count
+
+    def check_table_schema(
```
nit: maybe log_table_schema()? Otherwise it sounds like this is doing some validation

WillemKauf replied:
I wanted to add validation of some sort here, unfortunately it's a bit hard with a dynamically generated schema.

One could _maybe_ consider the successful call to `qe.run_query_fetch_all(f"describe {self.table_name}")` as validating, but i'm not opposed to changing the name.

andrwng commented on tests/rptest/tests/datalake/schema_scale_test.py:
```diff
@@ -0,0 +1,199 @@
+    @matrix(cloud_storage_type=supported_storage_types(),
+            query_engine=QUERY_ENGINES,
+            catalog_type=supported_catalog_types(),
+            cleanup_policy=[TopicSpec.CLEANUP_COMPACT],
```
Why is cleanup policy important for this test?

WillemKauf replied:
It isn't really important, I suppose this could be removed.

andrwng replied:
Why do we set it at all in the topic configs? It gets a default value, doesn't it?

WillemKauf replied:
I felt that `COMPACT_DELETE` is more testing coverage in terms of broker stress and subsystem interaction than just the default `DELETE`- let me know if you agree.

andrwng replied:
Ah, got it. I didn't realize that the goal was to stress CPU. Makes sense to me!

andrwng commented on tests/rptest/tests/datalake/schema_scale_test.py:
```diff
@@ -0,0 +1,199 @@
+                             query_engine: QueryEngineType,
+                             cleanup_policy: str, use_partition_spec: bool):
+        config = {
+            TopicSpec.PROPERTY_ICEBERG_INVALID_RECORD_ACTION: "dlq_table",
```
Why is DLQ important for this test? Should this test be checking that there is no DLQ table after completion?

WillemKauf replied:
Ah yeah, I was going to make that part of the validation step. Thanks for mentioning it

andrwng commented on tests/rptest/tests/datalake/schemas/data_types.py:
```diff
@@ -117,6 +117,26 @@ def to_proto() -> str:
     def random(enum_symbols) -> str:
         return str(random.choice(enum_symbols))
 
+    def make_valid_enum_name(n: int, producer_type: ProducerType) -> str:
```
nit: squash?

andrwng replied:
Also looks thike there is some protobuf change in this commit

andrwng commented on tests/rptest/tests/datalake/schema_scale_test.py:
```diff
@@ -184,7 +184,7 @@ def setUp(self):
             catalog_type=supported_catalog_types(),
             cleanup_policy=[TopicSpec.CLEANUP_COMPACT],
-            use_partition_spec=[False, True])
+            use_partition_spec=[False])
```
nit: squash?

andrwng reviewed (commented):
Looks great! Just a couple questions

WillemKauf committed:
`rptest`: add `schema_generator` and `data_type_generator`

For use in producing random protobuf/avro schemas.

Currently the implementation is fairly hardcoded for the avro spec;
future work will entail teasing this apart to leave the `data_type_generator`
and `schema_generator` fully generic, with further abstraction layers
on top to specialize for protobuf/avro schemas.

WillemKauf commented:
Force push to:
* Squash commits
* Add check for no DLQ after rounds of production using schema
* Remove cleanup policy parameterization
* Rename `check_table_schema()` to `log_table_schema()`

WillemKauf commented:
/ci-repeat 5
skip-redpanda-build
skip-units
dt-repeat=5
tests/rptest/tests/datalake/schema_scale_test.py

WillemKauf commented:
No flakes in [Build #64140](https://buildkite.com/redpanda/redpanda/builds/64140).

vbotbuildovich commented:
/backport v25.1.x
